ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java

```diff
@@ -795,7 +804,13 @@ public void alterTable(String catName, String dbName, String tblName, Table newT
       // Take a table snapshot and set it to newTbl.
       AcidUtils.TableSnapshot tableSnapshot = null;
       if (transactional) {
-        if (replWriteId > 0) {
+        if (AcidUtils.isTransactionalTable(newTbl) && !inReplication(newTbl) && replWriteId > 0) {
+          txnOpened = openTxnIfNeeded();
```

Comment:
Why do not we open the TXN at the same place where we do it for the normal queries?

Reply:
I did not get your comment. In DMLs as well, they check the value of "transactional" and then open a transaction within the same method.

Reply:
So txn is opened for every query in Compiler.analyze() - The we do not need to check the openTxns everywhere.
The writeId is requested in Driver.acquireLocks() - We might want to change the logic there, so writeId is requested for the table. I think it is much better to do these things in a single place than spreading out all over the code